RL.py

```diff
@@ -0,0 +1,274 @@
+        DONE = 0                              # Flag indicating if the episode has terminated
+        ep_return = 0                         # Initialize the return
+        step_counter = 0                      # Initialize the counter of episode steps
+        ep_arr = []
+
+        # START RL EPISODE
+        while True:
+            # Get current and next TO actions
+            action, next_TO_action = self.get_TO_actions(step_counter, action_TO=tau_TO)           
+
+            # Simulate actions and retrieve next state and compute reward
+            state_next, rwrd = env.step(rand_time, state_prev, action)
+            
+            # Store performed action and next state and reward
+            CACTO.control_arr = np.vstack([CACTO.control_arr, action.reshape(1,self.conf.nb_action)])
+            CACTO.state_arr = np.vstack([CACTO.state_arr, state_next.reshape(1,self.conf.nb_state)])
+            
+            CACTO.x_ee_arr.append(self.env.get_end_effector_position(CACTO.state_arr[-1,:])[0])
+            CACTO.y_ee_arr.append(self.env.get_end_effector_position(CACTO.state_arr[-1,:])[1])
+
+            ep_arr.append(rwrd)
+
+            if step_counter==CACTO.NSTEPS_SH-1:
+                DONE = 1
+
+            # Store transition if you want to use 1-step TD
+            if self.conf.TD_N==0:
+                prioritized_buffer.add(state_prev, next_TO_action, action, rwrd, state_next, float(DONE))
+
+            ep_return += rwrd                       # Increment the episodic return by the reward just recived
+            step_counter += 1                       
+            state_prev = np.copy(state_next)        # Next state becomes the prev state at the next episode step
+
+            # End episode
+            if DONE == 1:
+                break
+
+        # Plot the state and control trajectories of this episode 
+        # plot_results(tau0_arr,tau1_arr,tau2_arr,x_ee,y_ee,x_ee_arr,y_ee_arr,NSTEPS_SH,to=TO)
+        
+        # Store transition after computing the (partial) cost-to go when using n-step TD (up to Monte Carlo)
+        if self.conf.TD_N:
+            DONE = 0
+            cost_to_go_arr = []
+            for i in range(len(ep_arr)):          
+                final_i = min(i+self.conf.nsteps_TD_N,len(ep_arr))
+                if final_i == len(ep_arr):
+                    V_final = 0.0
+                else:
+                    if self.conf.NORMALIZE_INPUTS:
+                        state_next_rollout = CACTO.state_arr[final_i,:] / self.conf.state_norm_arr
+                        state_next_rollout[-1] = 2*state_next_rollout[-1] - 1
+                    else:
+                        state_next_rollout = CACTO.state_arr[final_i,:]
+                    tf_state_next_rollout = tf.expand_dims(tf.convert_to_tensor(state_next_rollout), 0)
+                    V_final = CACTO.target_critic(tf_state_next_rollout, training=False).numpy()[0][0]
+                cost_to_go = sum(ep_arr[i:final_i+1]) + V_final
```

Comment:
Unless I am wrong, this is a big mistake, which I thought we had already fixed in the past @gianluigigrandesso . What we store in the buffer should be the "partial cost-to-go", not the cost-to-go. This means we shouldn't use the Value network for computing this quantity, we should just sum the running costs from `i` to `i+nsteps_TD_N`